File: src/main/java/de/bixilon/minosoft/data/world/Chunk.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.world;

import de.bixilon.minosoft.data.entities.block.BlockEntityMetaData;
import de.bixilon.minosoft.data.mappings.blocks.Block;
import de.bixilon.minosoft.protocol.protocol.ProtocolDefinition;

import java.util.HashMap;

/**
 * Collection of 16 chunks sections
 */
public class Chunk {
    private final HashMap<Integer, ChunkSection> sections;

    public Chunk(HashMap<Integer, ChunkSection> sections) {
        this.sections = sections;
    }

    public Block getBlock(InChunkLocation location) {
        return getBlock(location.getX(), location.getY(), location.getZ());
    }

    public Block getBlock(int x, int y, int z) {
        int section = (y / ProtocolDefinition.SECTION_HEIGHT_Y);
        if (!this.sections.containsKey(section)) {
            return null;
        }
        return this.sections.get(section).getBlock(x, y % 16, z);
    }

    public void setBlock(int x, int y, int z, Block block) {
        int section = y / ProtocolDefinition.SECTION_HEIGHT_Y;
        createSection(section);
        this.sections.get(section).setBlock(x, y % 16, z, block);
    }

    void createSection(int height) {
        if (this.sections.get(height) == null) {
            // section was empty before, creating it
            this.sections.put(height, new ChunkSection());
        }
    }

    public void setBlocks(HashMap<InChunkLocation, Block> blocks) {
        blocks.forEach(this::setBlock);
    }

    public void setBlock(InChunkLocation location, Block block) {
        int section = (location.getY() / ProtocolDefinition.SECTION_HEIGHT_Y);
        createSection(section);
        this.sections.get(section).setBlock(location.getInChunkSectionLocation(), block);
    }

    public void setBlockEntityData(InChunkLocation position, BlockEntityMetaData data) {
        ChunkSection section = this.sections.get((position.getY() / ProtocolDefinition.SECTION_HEIGHT_Y));
        if (section == null) {
            return;
        }
        section.setBlockEntityData(position.getInChunkSectionLocation(), data);
    }

    public BlockEntityMetaData getBlockEntityData(InChunkLocation position) {
        ChunkSection section = this.sections.get((position.getY() / ProtocolDefinition.SECTION_HEIGHT_Y));
        if (section == null) {
            return null;
        }
        return section.getBlockEntityData(position.getInChunkSectionLocation());
    }

    public void setBlockEntityData(HashMap<InChunkLocation, BlockEntityMetaData> blockEntities) {
        blockEntities.forEach(this::setBlockEntityData);
    }

    public HashMap<Integer, ChunkSection> getSections() {
        return this.sections;
    }

    public ChunkSection getSectionOrCreate(int sectionHeight) {
        ChunkSection section = this.sections.get(sectionHeight);
        if (section == null) {
            section = new ChunkSection();
            this.sections.put(sectionHeight, section);
        }
        return section;
    }
}

File: src/main/java/de/bixilon/minosoft/data/world/chunk/chunk/Chunk.kt
/*
 * Minosoft
 * Copyright (C) 2020-2026 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */
package de.bixilon.minosoft.data.world.chunk.chunk

import de.bixilon.kutil.cast.CastUtil.unsafeNull
import de.bixilon.kutil.concurrent.lock.LockUtil.acquired
import de.bixilon.kutil.concurrent.lock.LockUtil.locked
import de.bixilon.kutil.concurrent.lock.locks.reentrant.ReentrantRWLock
import de.bixilon.minosoft.data.Tickable
import de.bixilon.minosoft.data.entities.block.BlockEntity
import de.bixilon.minosoft.data.registries.blocks.state.BlockState
import de.bixilon.minosoft.data.world.World
import de.bixilon.minosoft.data.world.biome.source.BiomeSource
import de.bixilon.minosoft.data.world.chunk.ChunkSection
import de.bixilon.minosoft.data.world.chunk.light.section.ChunkLight
import de.bixilon.minosoft.data.world.chunk.neighbours.ChunkNeighbours
import de.bixilon.minosoft.data.world.chunk.update.block.ChunkLocalBlockUpdate
import de.bixilon.minosoft.data.world.chunk.update.block.ProposedBlockChange
import de.bixilon.minosoft.data.world.chunk.update.chunk.ChunkLightUpdate.Causes
import de.bixilon.minosoft.data.world.positions.ChunkPosition
import de.bixilon.minosoft.data.world.positions.InChunkPosition
import de.bixilon.minosoft.data.world.positions.SectionHeight

/**
 * Collection of chunk sections (height aligned)
 */
class Chunk(
    val world: World,
    val position: ChunkPosition,
) : Tickable {
    val lock = ReentrantRWLock()
    val light = ChunkLight(this)
    val sections = ChunkSectionManagement(this)
    var biomeSource: BiomeSource? = null

    val neighbours = ChunkNeighbours(this)

    operator fun get(sectionHeight: SectionHeight): ChunkSection? = sections[sectionHeight]

    operator fun get(position: InChunkPosition): BlockState? {
        return this[position.sectionHeight]?.blocks?.get(position.inSectionPosition)
    }

    operator fun set(position: InChunkPosition, state: BlockState?) {
        sections.create(position.sectionHeight)?.set(position.inSectionPosition, state)
    }

    fun getBlockEntity(position: InChunkPosition): BlockEntity? {
        return this[position.sectionHeight]?.entities?.get(position.inSectionPosition)
    }

    fun updateBlockEntity(position: InChunkPosition): BlockEntity? {
        return this[position.sectionHeight]?.entities?.update(position.inSectionPosition)
    }

    fun apply(update: ProposedBlockChange) {
        this[update.position] = update.state
    }

    private fun unsafeApply(vararg updates: ProposedBlockChange): Set<ChunkLocalBlockUpdate.Change> {
        val executed: MutableSet<ChunkLocalBlockUpdate.Change> = HashSet(updates.size)

        for (update in updates) {
            val (position, state) = update
            val sectionHeight = position.sectionHeight

            var section = this[sectionHeight]
            if (state == null && section == null) continue

            section = this.sections.create(sectionHeight) ?: continue

            val previous = section.blocks.set(position.inSectionPosition, state)

            if (previous == update.state) continue
            if (previous?.block != state?.block) {
                section.entities[position.inSectionPosition] = null
            }

            section.entities.update(position.inSectionPosition)

            executed += ChunkLocalBlockUpdate.Change(update.position, previous, update.state)
        }

        return executed
    }

    fun apply(vararg updates: ProposedBlockChange) {
        if (updates.isEmpty()) return
        if (updates.size == 1) return apply(updates.first())


        var executed: Set<ChunkLocalBlockUpdate.Change> = unsafeNull()
        lock.locked {
            executed = unsafeApply(*updates)

            if (executed.isEmpty()) return

            light.heightmap.recalculate() // TODO: Only changed ones
            light.recalculate(fireEvent = false, cause = Causes.INITIAL)
        }

        if (neighbours.complete) {
            val sections: HashSet<ChunkSection> = hashSetOf()

            for (change in executed) {
                sections += this[change.position.sectionHeight] ?: continue
            }

            for (section in sections) {
                light.fireLightChange(section, Causes.BLOCK_CHANGE)
            }
            light.fireLightChange(Causes.PROPAGATION)
        }

        ChunkLocalBlockUpdate(this, executed).fire(world.session)
    }

    @Deprecated("sections.create")
    fun getOrPut(height: Int, light: Boolean = true) = sections.create(height, light)

    override fun tick() {
        if (!neighbours.complete) return
        lock.acquired { sections.forEach { it.tick() } }
    }

    override fun toString() = "Chunk($position)"
}


